[#11] 주문생성 시 재고감소, 배달객체 생성을 Kafka로 처리

yooyouny opened the pull request:
### 🌱 작업 사항 
주문생성요청
1. Order, OrderDetail 생성
2. 상품 재고 감소 (Kafka로 비동기 처리)
3. Order 정보로 Delivery 객체 생성 (Kafka로 비동기 처리)
(배달경로생성은 Order의 상태를 REQUESTED로 바뀔때 생성) 

### ❓ 리뷰 포인트
- 상품재고감소를 하려고 Company 서버에서 ProductDeductConsumer 생성했고 테스트 진행을 위해 user가 ddl-auto create로 되어있어서 update로 수정하는 등 다른 도메인에서 수정한 부분들이 많아서 혹시 잘못된 부분이 있는지 확인 부탁드립니다
- 재고감소 실패시 에러토픽을 발행해서 주문상태를 취소처리하고 비동기로 생성한 Delivery 객체를 soft delete 합니다
- 배송객체생성 실패시 에러토픽을 발행해서 주문상태를 취소처리하고 비동기로 처리한 재고감소 로직을 revert합니다

- 테스트 첨부
<img width="846" alt="image" src="https://github.com/user-attachments/assets/6faeb634-3210-4c79-a449-1c2fbcebbf09">
주문생성
<img width="1168" alt="image" src="https://github.com/user-attachments/assets/f90c4ded-048d-4f70-954f-bc15cd8f12c1">
재고감소
<img width="1180" alt="image" src="https://github.com/user-attachments/assets/6337892e-e320-4520-a5c7-3e9439efbf36">
배달객체생성


### 🦄 관련 이슈
resolves #11

yooyouny committed:
add: 주문 생성 API 추가 (#11)
- Order, OrderDetail 객체 생성
- 총금액, 총수량 연산
- 배송, 배송루트 생성은 연관허브정보가 정해지면 추가
- 배송, 배송루트는 현재 주문생성에 반드시 반환해야할 요소가 아니기 때문에 비동기로 생성할 예정

yooyouny committed:
fix: 공통객체의 code에 상태코드 이름을 넣을 수 있도록 생성자 수정

yooyouny committed:
fix: 성공응답 객체 생성시 data만 넣는것으로 변경

yooyouny committed:
add: 주문 생성 API에 재고감소 비동기 처리 로직 추가 #11
- kafka 설정
- kafka로 send하는 OrderProductDeductProducer 등록
- 재고감소에 대한 에러 처리는 product 개발이 완료된 후 진행할 예정

yooyouny committed:
rename: Delivery 관련 공통 dto들 정의

yooyouny committed:
fix: company의 application name으로 떠서 삭제

yooyouny committed:
chore: gateway에 delivery 라우팅 설정

yooyouny committed:
fix: user의 ddl-auto 방식을 update로 변경

yooyouny committed:
chore: 카프카 직렬화, 역직렬화 설정 수정

yooyouny committed:
add: Company에 재고감소 카프카리스너 등록 및 재고감소 로직 추가
- 카프카 의존성 추가

yooyouny committed:
add: 배달생성의 에러핸들링을 위한 컨슈머 등록
- 배달서버에서 배달객체 생성 실패시 해당 주문의 상태를 취소로 변경

yooyouny committed:
add: 공통객체에 kafka의 토픽들을 상수화한 객체 생성

yooyouny committed:
add: 주문 생성 시 delivery 서버로 객체생성을 비동기로 요청

yooyouny committed:
add: 배달객체생성 요청을 받을 컨슈머 등록
- companyId로 hubClient를 통해 HubId를 가져오는 feign 요청 추가

yooyouny committed:
fix: 허브경로를 가져오는 hubClient의 name을 application.name 설정으로 변경

gidopa commented on service/delivery/delivery_dto/src/main/java/com/sparta/delivery/dto/DeliveryRouteDto.java:
```diff
@@ -0,0 +1,34 @@
+  private UUID arrivalHubId;
+  private double estimatedDistance;
+  private long estimatedElapsedTime;
+  private Integer realDistance;
```
estimatedDistance는 double인데 realDistance는 Integer인 이유가 혹시 따로 있을까요 ?!
그리고 혹시 realDistance 는 실제 경로 추적을 못하는 상황에서 어떻게 만들지 궁금합니다 !

yooyouny replied:
테이블 명세서에 정의한것과 현재 사용하지 않는 컬럼들 정리하겠습니다!

gidopa reviewed (approved):
주말에 고생 많으셨습니다 !

kimgunwooo commented on service/delivery/delivery_server/src/main/java/com/sparta/delivery/infrastructure/messaging/DeliveryCreateConsumer.java:
```diff
@@ -0,0 +1,35 @@
+  private final DeliveryFacadeService deliveryFacadeService;
+  private final KafkaTemplate<String, Object> kafkaTemplate;
+
+  @KafkaListener(topics = KafkaTopicConstant.CREATE_DELIVERY, groupId = "delivery")
+  public void consume(
+      @Header(name = "kafka_receivedMessageKey") String key, DeliveryCreateDto deliveryCreateDto) {
```
hashmap으로 받아서 dto로 parse 하는 부분도 있고, dto 그대로 받는 부분도 있는데, 나눈 이유가 있을까요??

yooyouny replied:
kafka의 value들은 전부 JSON타입으로 직렬화하는데 dto를 리스트로 감싸니까 바로 dto로 역직렬화하지는 못하더라고요 그래서 제네릭 타입을 ProductDeductDto.class 등으로 명시해주는 식으로 받으니 에러가 안나서 리스트타입으로 수신하는것만 그렇게 처리했습니다

yooyouny committed:
add: 재고감소오류시 주문을 취소처리하고 만들었던 배달엔티티삭제하는 로직 추가
- company에서 재고감소 소비중 에러가 발생하면 ERROR_IN_DEDUCT_PRODUCT_QUANTITY 발생
- order가 해당 토픽을 수신하고 있다가 감지하면 order 취소처리, DELIETE_DELIVERY 발생
- 해당 토픽을 수신하고 있던 DELIVERY가 softDelete 처리완료

yooyouny committed:
add: 배달생성오류시 주문을 취소처리하고 감소했던 재고를 원복하기
- delivery에서 배달생성오류가 발생하면 ERROR_IN_CREATE_DELIVERY 발생
- order가 해당 토픽을 수신하고 있다가 감지하면 order 취소처리, REVERT_PRODUCT_QUANTITY 발생
- 해당 토픽을 수신하고 있던 company가 받은 productId들과 quantity를 원복
